feat(fs): add UNC paths support in fs.normalize

pysan3 opened the pull request:
As stated here https://github.com/neovim/neovim/issues/27068, Windows network storage paths breaks with `fs.normalize`.

- `\\wsl$\Ubuntu\home\user`
- `\\127.0.0.1\path`

> I do need the information that the path starts with a double slash, but I want to evaluate that after `expand_env`, so I can't just check it against the path before running this function as well.
> 
> Could you add an option something like `opts.keep_dup_slashes (default: false)` that does not delete the duplicated slashes, and return the slashes as-is? (don't `:gsub('/+', '/')`).
> 
> I'd love to hear your opinions @lewis6991 @gpanders .
> 
> BTW, I don't have problem with `gsub('/+$', '')` (I hope this regex is correct) to delete the trailing slash.

cc @lewis6991 @gpanders 

### Tests

```lua
    it('works with / (opts.unique_slash = false)', function()
      eq('///', vim.fs.normalize('///', { unique_slash = false }))
    end)
    it('removes duplicated slashes (opts.unique_slash = true)', function()
      eq('/home/user', vim.fs.normalize('//home//user//'))
    end)
    it('do not remove duplicated slashes (opts.unique_slash = false)', function()
      eq('//home//user', vim.fs.normalize('//home//user//', { unique_slash = false })) -- does remove trailing slashes
    end)
```

Closes #27068

### PS

This is my first PR to neovim repo. Please kindly give me instructions if I have something missing.

pysan3 committed:
feat(fs): add collapse_slash to fs.normalize to keep duplicate slashes

wookayin commented on runtime/lua/vim/fs.lua:
```diff
@@ -339,13 +339,15 @@ end
 ---@param path (string) Path to normalize
 ---@param opts table|nil Options:
 ---             - expand_env: boolean Expand environment variables (default: true)
+---             - unique_slash: boolean Remove duplicated slashes (default: true)
```
```suggestion
---             - expand_env: (boolean) Expand environment variables (default: true)
---             - unique_slash: (boolean) Remove duplicated slashes (default: true)
```

pysan3 replied:
Thanks. The `expand_env` line is not a line I added but can I edit this line in my PR (or is it not a big deal)?

pysan3 replied:
Fixed in [`0a3876a` (#27202)](https://github.com/neovim/neovim/pull/27202/commits/0a3876aa9852671900bff35f0a35dcbc5db43a38)

wookayin commented on runtime/doc/lua.txt:
```diff
@@ -2926,6 +2926,8 @@ vim.fs.normalize({path}, {opts})                          *vim.fs.normalize()*
       • {opts}  (`table?`) Options:
                 • expand_env: boolean Expand environment variables (default:
                   true)
+                • unique_slash: boolean Remove duplicated slashes (default:
```
The parameter name `unique_slash` sounds a bit awkward. I can suggest:
- `normalize_slash` (or `normalize_slashes`)
- `collapse_slash` (or `collapse_slashes`)

Normalize slash would be a more common term used in other programming languages' stdlib.

pysan3 replied:
Nice one. I like `collapse_slash`.

justinmk commented on test/functional/lua/fs_spec.lua:
```diff
@@ -291,9 +291,18 @@ describe('vim.fs', function()
+      eq('/home/user', vim.fs.normalize('//home//user//'))
+    end)
+    it('do not remove duplicated slashes (opts.collapse_slash = false)', function()
+      eq('//home//user', vim.fs.normalize('//home//user//', { collapse_slash = false })) -- does remove trailing slashes
```
This makes no sense. If `normalize()` doesn't trim the last slashes, like this:

    //home/user/

then its result won't be the same for these inputs:

    //home//user//
    //home/user/

which defeats the purpose of normalizing. The normalized result should always be the same.

pysan3 replied:
It **does** remove any amount of trailing slashes, and keeps any duplicated slashes before the end of the path. Does that make sense?

justinmk replied:
No. `//home//user//` and `//home/user/` are equivalent paths, yet `normalize({collapse_slash=false})` would return different results. Why can't it return `//home/user` if the input is `//home//user//`?

pysan3 replied:
As the literal option name suggests, this PR was intended to toggle whether to collapse multiple slashes into one. 

I didn't want to talk about any particular file system but more from a programming perspective and stop `gsub("/+", ...)` since it is an irreversible operation. 

I think fs.normalize has three missions. 

1. Change all `\` to `/`
2. Expand envvars (eg $HOME)
3. Shrink duplicated slashes. 

As we do have an option to toggle `2.`, I thought why couldn't there be an option to toggle `3.`?

justinmk replied:
> I think fs.normalize has three missions.

No, its mission is to produce a canonical result for all equivalent forms of any kind of path.

pysan3 replied:
This is a nice idea. Added!

pysan3 replied:
@dundergoc Capitalize drive name.

justinmk reviewed (changes requested):
In the worst case, maybe it makes sense to add a `unc: boolean` parameter. So that `normalize()` understands that the first `//` should be preserved, but nothing after that.

Alternatively, can we just _recognize_ a UNC path? I.e. if path starts with `//` treat it as UNC (on Windows only).

pysan3 commented:
Do you think unc = true should be the default?

justinmk commented:
# Proposal

- Let's start by special-casing `//foo`. Although that violates the "determinism" of `normalize()`, in practice it's unlikely to matter. 
    - If I'm wrong about that, we can always add `unc:boolean` later.

So the following test cases represent what I propose:

```
//foo => //foo      # UNC path.
///foo => //foo     # UNC path. Alternatively we could treat as invalid UNC, then => /foo
/foo => /foo
//foo//bar//// => //foo/bar
```

pysan3 committed:
feat(fs): add UNC paths support in fs.normalize

pysan3 commented:
Thank you very much for your feedbacks.

I totally agree with @justinmk 's proposal, and fixed the code to support UNC paths.

I've removed `collapse_slashes` option and changed the code to support UNC paths out of the box.
I've also added test cases listed in https://github.com/neovim/neovim/pull/27202#issuecomment-1919183890 in `fs_spec.lua`. Thanks @justinmk !

As suggested [here](https://github.com/neovim/neovim/pull/27202#discussion_r1472897840), I also made it so drive name in Windows paths (`C:/`) will always be capitalized (and added tests as well).

I'd be more than happy to hear your opinions! @justinmk @wookayin @lewis6991 @gpanders @dundargoc 

P.S Are the comments useful? I'm not a native English speaker so the explanation might not be clear.

pysan3 committed:
docs(fs): remove deleted option in fs.normalize

wookayin commented on test/functional/lua/fs_spec.lua:
```diff
@@ -307,10 +313,26 @@ describe('vim.fs', function()
         )
       )
     end)
+    it('works with UNC paths', function()
+      -- See https://github.com/neovim/neovim/issues/27068 for more detail
```
```suggestion
    it('works with UNC paths #27068 #27202', function()
```

wookayin commented on test/functional/lua/fs_spec.lua:
```diff
@@ -285,12 +285,18 @@ describe('vim.fs', function()
       eq('C:/Users/jdoe', vim.fs.normalize('C:\\Users\\jdoe'))
     end)
+    it('works with mixed forward and backward slashes', function()
+      eq('C:/Users/jdoe', vim.fs.normalize('C:\\//Users//\\jdoe//\\'))
```
The single-letter string represented as `\\` instead of `\` due to escaping is quite confusing given that we now want to deal with double slashes. Can you change `'...\\...'` with `[[...\...]]` (including existing test cases)? This will make tests more readable.

pysan3 replied:
Fair enough. I'll change all of them into raw string notation.

pysan3 replied:
@dundargoc

pysan3 committed:
doc(fs): fix type notation

Co-authored-by: Jongwook Choi <wookayin@gmail.com>

pysan3 committed:
Update test name in lua/fs_spec.lua

Co-authored-by: Jongwook Choi <wookayin@gmail.com>

pysan3 committed:
test(fs_spec): use raw string notation when path contains backslash

lewis6991 commented on runtime/lua/vim/fs.lua:
```diff
@@ -349,22 +349,24 @@ function M.normalize(path, opts)
+  -- `//` is not a valid UNC path, hence converted to `/` with `:gsub('/$', '')`.
+  -- This also keeps traling slash in the case of root `/`, but removes it otherwise.
+
+  if iswin and path:sub(2, 2) == ':' then -- capitalize windows drive name
```
Should probably check the first char is `[a-zA-Z]`

pysan3 replied:
I don't think so. Colon is not a valid character for a path string except right after the device name.

And even if it is true, I'm just doing `string.upper(first)` which does nothing when first is not an alphabet in the first place.

lewis6991 replied:
This function needs to gracefully handle any old junk that's passed to it, so it needs to be sensible with invalid paths. So we should only mangle if the path matches `^[a-zA-Z]:/`, otherwise we just leave it alone.

pysan3 replied:
I mean that `string.upper` will do nothing if the first letter is not an alphabet. Am I right?

pysan3 replied:
@lewis6991 Do I need to change anything here?

lewis6991 replied:
Yes, this code assumes `path[3] == '/'` which it might not be. I think we need to match `^[a-zA-Z]:/` exactly here so we only mangle if we absolutely need to.

pysan3 replied:
No. Due to `path:sub(2):...:gsub('/$', '')`, `C:/` will be converted to `C:` temporary and a slash must be added afterwards to negate this behavior.

Please read the code.

lewis6991 replied:
> Please read the code.

Please be polite. I don't have to review this.

> Due to path:sub(2):...:gsub('/$', ''), C:/ will be converted to C: temporary and a slash must be added afterwards to negate this behavior.

This is not obvious at all. And despite this, it still assumes `path[3] == '/'` for cases other than `C:/`

pysan3 replied:
I thought you understood this while writing https://github.com/neovim/neovim/pull/27202#discussion_r1474262585? 
It is the exact behavior you wrote in the comments.

> And despite this, it still assumes `path[3] == '/'` for cases other than `C:/`

In [this issue](https://github.com/neovim/neovim/issues/27068) you refused to bring UNC paths because it is not supported by the Unix path standard. And now I feel like you are trying to support all edge cases for every file system. May I ask what direction you are towards right now?

Besides that, is there an edge case where `iswin and path:sub(2, 2) == ':'` but `path[3] ~= '/'`?

lewis6991 replied:
> I thought you understood this while writing https://github.com/neovim/neovin/pull/27202#discussion_r1474262585?
It is the exact behavior you wrote in the comments.

No, not at all. That specific code is all about normalizing slashes AND preserving leading `//`. It has nothing to do with the single `C:/` edge case.

> In https://github.com/neovim/neovim/issues/27068 you refused to bring UNC paths because it is not supported by the Unix path standard. And now I feel like you are trying to support all edge cases for every file system. May I ask what direction you are towards right now?

We support it or we don't support it. We've decided to support it and there is a glaring bug in this code.

To put it simply.

```
C:some_invalid_path
```
should NOT normalize to:
```
C:/ome_invalid_path
```

pysan3 replied:
Thank you for your input. I find your words quite hurtful, so I would prefer not to contribute anymore.

I wish the best of luck for this project.

lewis6991 commented on runtime/lua/vim/fs.lua:
```diff
@@ -349,22 +349,24 @@ function M.normalize(path, opts)
-  path = path:gsub('\\', '/'):gsub('/+', '/')
-  if iswin and path:match('^%w:/$') then
-    return path
+  local first = path:sub(1, 1)
+  path = (first == '\\' and '/' or first) .. path:sub(2):gsub('[\\/]+', '/'):gsub('/$', '')
```
Shouldn't the `\` -> `/` be unconditional?

```suggestion
  path = path:gsub('\\', '/')
  path = path:sub(1, 1) .. path:sub(2):gsub('/+', '/')
```

pysan3 replied:
This is wrong. We must have `:gsub('/$', '')` to strip the trailing slash (eg `foo/bar/`)

Your solution will then require to call `gsub` 3 times, which is not efficient.

Iiuc, we should assume this function will be called _a lot_ of times, so we should take performance over code readability, and tbh, I think my solution is quite clean ngl.

lewis6991 replied:
If you want to avoid another gsub then I suggest not doing `:gsub('/$', '')` and instead do 

```lua
if path:sub(-1) == '/' then
  path = path:sub(1, -2)
end
```

> so we should take performance over code readability, and tbh, I think my solution is quite clean ngl.

Nothing indicates this function is a bottleneck in any code path. And if you're dealing with the file system any `gsub`'s are going to be completely dwarfed by calls out to the system.

pysan3 replied:
I don't think this is clean and I think it makes the code less readable. So I see no reason to change my code.

> Nothing indicates this function is a bottleneck in any code path. And if you're dealing with the file system any `gsub`'s are going to be completely dwarfed by calls out to the system.

I see.

lewis6991 replied:
That's unfortunate. Personally, I don't think squashing many terms into a single line that do multiple things is clean at all. It's short, but it isn't clean or easy to understand. By breaking it down into multiple independent operations, the reader can comprehend each distinct step separately.

And if we keep the additional `gsub`, because realistically this isn't going meaningfully affect performance, then I fail to see how:

```lua
local first = path:sub(1, 1)
path = (first == '\\' and '/' or first) .. path:sub(2):gsub('[\\/]+', '/'):gsub('/$', '')
```

Is cleaner than:
```lua
path = path:gsub('\\', '/')  -- Normalize slashes
path = path:sub(1, 1) .. path:sub(2):gsub('/+', '/'):gsub('/$', '')  -- Remove redundant slashes
```

pysan3 replied:
I don't think your code explains it well enough because _Remove redundant slashes_ does not explain `sub(1, 1)` and its quite unclear anyways.

It doesn't document that `//` will result in `/` but `//foo` will be `//foo`, and `///foo` will be `//foo`.

I didn't want to say that my code is cleaner, but I meant any code will not be clear anyways, and didn't find a reason to introduce another gsub for little benifit.

pysan3 committed:
fix(fs): add gsub to satisfy the reader

lewis6991 commented on runtime/lua/vim/fs.lua:
```diff
@@ -349,22 +349,26 @@ function M.normalize(path, opts)
+  -- `//` is not a valid UNC path, hence converted to `/` with `:gsub('/$', '')`.
+  -- This also keeps traling slash in the case of root `/`, but removes it otherwise.
+
+  if iswin and path:sub(2, 2) == ':' then -- Capitalize Windows drive name
```
```suggestion
  if iswin and path:match('^[a-zA-Z]:/') then -- Capitalize Windows drive name
```
